language-plutus-core/src/Language/PlutusCore/Core/Instance/Eq.hs

```diff
@@ -52,17 +52,148 @@ instance (HasUniques (Program tyname name uni ann), GEq uni, Closed uni, uni `Ev
 
 type EqRenameOf ren a = HasUniques a => a -> a -> EqRename ren
 
+{- Note [Eta-equality]
+During type checking we may end up performing an equality check of two types that looks like this:
+
+    h f == h (\x -> f x)
+
+In particular, this is likely to happen when type checking functions operating on lambda-encoded
+GADTs (like in the @.../Examples/Data/Vec.hs@ file). Such an equality check amounts to checking
+that functions from both the sides of the equation are equal (both the functions are @h@, so this
+checks succeeds) and checking that arguments are equal as well. I.e.
+
+    f == (\x -> f x)
+
+has to hold. But whether it holds or not depends on whether we support eta-equality for type-level
+functions or not (see https://ncatlab.org/nlab/show/eta-conversion).
+
+It might seem that the simplest way of supporting eta-equality is by eta-expanding the side of the
+equation that doesn't start with a lambda (the LHS in the example above). That would transform our
+example into
+
+    (\x -> f x) == (\x -> f x)
+
+which can be checked with our usual strategy from there on. However, we cannot simply copy the
+lambda-binding from the RHS, because the variable that it binds can be referenced (as a free
+variable) in @f@ from the LHS, in which case naively prepending a lambda would change the meaning
+of the type on the LHS. For example if we have the following equality check
+
+    f x == (\x -> f x x)
+
+then we cannot simply expand the LHS to @\x -> f x x@ by copying the lambda from the RHS as that
+would change the meaning of the type.
+
+So what we are forced to do is to invent a fresh variable, lambda-bind it and apply the function
+to that variable to transform the original example into
+
+    (\y -> f y) == (\x -> f x)
+
+where @y@ is not referenced in @f@. Checking equality of these two types is still a regular thing.
+However, in order to invent a fresh variable we need to be in the 'Quote' monad and we need to know
+what free variables both the sides of the equation reference. I.e. we have to perform a prepass on
+both the sides before checking their equality. This is a rather big cost to pay for such a simple
+feature, especially given the fact that the two sides may start with different type constructors and
+we want to return 'False' immediately in that case instead of preprocessing anything fully (we could
+try using some meta-circular programming to overcome this inefficiency problem, but that'd be a
+terrible overkill).
+
+What we're left with is eta-contracting the side of the equation that starts with a lambda. I.e.
+turning the original equation into
+
+    f == f
+
+by eta-contracting @\x -> f x@ (the RHS) into @f@. However, we cannot cut @\x -> f x@ into @f@ and
+proceed checking equality without doing anything about @x@. Consider:
+
+    f x == (\x -> f x x)
+
+we can't simply cut @(\x -> f x x)@ into @f x@, because on the LHS we have a type that references
```

Comment:
I don't quite understand this. I thought the standard eta-contraction rule was something like:
`if x \not \in FV(f) then \x . f x == f`. Isn't that enough here? i.e. we should check if `x` is free in `f x`, find that it is, and so say we can't eta-contract it.

We would need to be careful about not doing lots of wasteful free-variable calculations, and maybe what you're suggesting is something like a dual version of this, where we record variables that must not be free and then fail later. But it seems like the free-variable version would be a lot simpler, at least.

Reply:
> We would need to be careful about not doing lots of wasteful free-variable calculations, and maybe what you're suggesting is something like a dual version of this, where we record variables that must not be free and then fail later. But it seems like the free-variable version would be a lot simpler, at least.

And a lot less efficient, especially in a case where we have two very different types, one starts with a lambda, the other does not and we start traversing the body of that lambda to see if eta-contraction is possible instead of stripping the lambda (an O(1) thing) and failing immediately once we see that two leading type constructors do not match.

Reply:
*waves hand* doesn't laziness help? :p 

But sure, this is probably expensive. I don't know if it's too expensive, though.

Maybe we should have a `dumbEtaContract` implementation that we can test this against, though.

Reply:
> waves hand doesn't laziness help? :p

Laziness doesn't save us from fully processing the body of the lambda to check if eta-contraction is possible before we can look at the next head.

> Maybe we should have a dumbEtaContract implementation that we can test this against, though.

The problem is that we need to generate interesting enough terms for that. And if we can do that, then we can just look at generated terms and predict if one is an eta-contracted version of the other. That's what I do in the tests.